### ISSUE TITLE
Mac tooltips associate with one desktop

### ISSUE THREAD

**Issue by ogallagher** [fontforge/fontforge#4927]
## Environment

FontForge **master** (commit [0bf72b2f](https://github.com/fontforge/fontforge/commit/0bf72b2f8a9ba5c784e23586c5b8dfc0658d9d89))

Mac OS Catalina 10.15.7

## Virtual desktops

I use [virtual desktops/spaces](https://support.apple.com/en-us/HT204100) regularly. When I open fontforge, the font viewer window will open in my current desktop.

## Problem behavior

If I open the glyph editor for one of the characters, and move the glyph editor window to a different desktop, the **tooltips will still display in the original desktop** where I first launched fontforge. This same behavior happens if I move any of the fontforge windows to a different desktop, including the font viewer.

https://user-images.githubusercontent.com/17031438/156927408-83e29355-a297-47d7-908d-fb39ac84502f.mp4

## Desired behavior

A tooltip should be associated with whatever window (and desktop) it came from, so that a window in desktop 1 will show a tooltip in desktop 1, and a window in desktop 2 shows a tooltip in desktop 2.

**Comment by JoesCat** [comment]
As simple a request as this appears to be at the visual level, it is actually a very big request, and the problems created outweigh the benefits at a programming level, user level, and security level too.

The simplest solution I would suggest, is to open each Fontforge with a different colour theme. Example, let's say, green, yellow, red. This way, you can tell them apart from each other.

From a programming point of view, graphical programs tend to work as a parent->child->child->child type of concept. In terms of Fontforge, the core/parent is the initial window that opens up when the splash screen appears. Then the child processes are all the windows you open from this core/parent process. If you open another fontforge, again, the core/parent for that starts at about the splash screen, opeing up branch child processes (like branches on a tree).
For a developer to change this behaviour for Fontforge, involves a lot of re-writing of code. This can also create and add new bugs in what is otherwise stable code.
Since Fontforge is maintained by volunteers, you are asking them to put a lot of free time into something that could have been used elsewhere with more benefit to everybody.

From a user point of view, a lot of users are already used to this sort of behavior, and to deviate from it will cause problems for people that are already expecting the current behaviour. There are even users that still use Fontforge from 2012, because they expect certain behaviour that worked then, and is different now. Also, many users tend to expect this sort of parent/child program behaviour since it is considered "the norm", for many graphical programs other than Fontforge.
There are other scenarios where the current behaviour has advantage, and I do this often in comparing one program with another. Other people will have their reasons too.

From a security point of view, let's use your virtual sandbox as an example. Let's say that we have a rogue/bad program that infects other programs. As a parent/child type process, if you were to slide one of the windows from a bad program over your desktop, it could potentially lock onto a good program on your clean desktop, and infect your program, and that program can infect other programs after that. As parent/child behaviour, the only program that the "bad" window could infect, is it's parent process, therefore, nothing really available to grab and infect.

**Comment by jtanx** [comment]
From someone who's actually familiar with this: it's a gdk problem that again goes away when moving to Qt, so yeah it's fixable but it'll probably be a while before that eventuates.

**Comment by jtanx** [comment]
I mean anything's possible if you spend enough time at it, but given I've already attempted a GTK port and gave up (#3295), I can see a clearer way forward with Qt that doesn't involve crippling limitations, massive rewrites, or having to constantly deal with bugs when not running on Gnome/Linux.

**Comment by skef** [comment]
I had thought we were at a point where you had gone quite a bit down the path of a GTK port that wound up looking like a lot of work/change, and QT mostly looked favorable for big-picture reasons (like better cross-platform support). I wasn't aware you had gone far enough down the QT path to be able to judge the specific work involved. Have you now gotten farther with QT than you did with GTK? Did QT finally get c-bindings of some kind? 

When I looked at your GTK branch and thought about GTK 4 it looked like the big obstacle was event handling, and that could be improved by moving gdraw itself to a glib-based event model. That would indeed be a bunch of work but it's not out of the question. I doubt dealing with the window positioning stuff would wind up being a huge deal.

**Comment by jtanx** [comment]
> Have you now gotten farther with QT than you did with GTK? Did QT finally get c-bindings of some kind?

Yeah, I'd say I'm pretty far-advanced into it, and like I mentioned in #3295, I've gotten further than I ever did with the GTK port, and things 'just work'. No, it's written in C++ but that's basically constrained to one file only; I'm just implementing the gdraw interface after all.

**Comment by skef** [comment]
Oh, weird, for some reason GitHub didn't send any of those updates to my email, even though I've commented on the issue. 

Interesting. If it works that's great!
